Include the original `LineNumberNode`s from the original expression in the final expression

DilumAluthge opened the pull request:
Fixes #178

chriselrod commented:
The ideal solution would probably try and place line number nodes next to somewhat related code. If that happens, I think users should get slightly more helpful error messages when something goes wrong pointing towards an associated line in their own code. 
However,
1. That'd be substantially more complicated to implement, and should wait for the eventual overhaul I've been promising for months.
2. That'd probably be a lot less helpful than it sounds. If you're debugging a problem, the first thing I'd try is removing `@avx` to confirm that the code runs correctly without it. If it doesn't, then fix it before adding the `@avx` (and obviously what `@avx` does won't matter then while you're fixing it). On the other hand, if it does run correctly without `@avx` but incorrectly with it, I'm not sure how useful this would be in diagnosing bugs in the macro; my concern is it could produce red herrings or add to confusion.

So for now, just dumping all of them at the start of the macro seems fine.
I'll merge and make a release after tests pass (mind bumping the version?), will be good to confirm this works and see what this does to reported coverage of libraries making use of LoopVectorization.

DilumAluthge commented:
Yeah, I do think it would be better for the line number nodes to go right next to the associated code.

DilumAluthge commented:
Here are the instructions for running a short demo. Note that the script
will delete your `~/.julia` directory, so you may want to back it up first.

Create a file named `code.jl` with the following contents:
```julia
using LoopVectorization

function f!(x::AbstractVector)
    @avx for i ∈ axes(x, 1)
        x[i] = 1
    end
    return x
end
```

Create a file named `test.jl` with the following contents:
```julia
include("code.jl")

using Test

x = zeros(Int, 10)

@test all(x .== 0)

f!(x)

@test all(x .== 1)
```

Create a file named `run.sh` with the following contents:
```bash
#!/bin/bash

rm -rf ~/.julia
rm -rf *.cov
julia -e 'import Pkg; Pkg.add(Pkg.PackageSpec(url = "https://github.com/chriselrod/LoopVectorization.jl", rev = "master"))'
julia --code-coverage test.jl
ls *.cov
cat -n *.cov

rm -rf ~/.julia
rm -rf *.cov
julia -e 'import Pkg; Pkg.add(Pkg.PackageSpec(url = "https://github.com/DilumAluthge/LoopVectorization.jl", rev = "dpa/avx-coverage"))'
julia --code-coverage test.jl
ls *.cov
cat -n *.cov

rm -rf ~/.julia
rm -rf *.cov
```

Make sure that the `run.sh` script is executable by running `chmod +x run.sh`.

Now run the script: `./run.sh`.

Here is the output that I receive:
```
$ ./run.sh
     Cloning git-repo `https://github.com/chriselrod/LoopVectorization.jl`
    Updating git-repo `https://github.com/chriselrod/LoopVectorization.jl`
   Resolving package versions...
  Installing known registries into `~/.julia`
     Cloning registry from "https://github.com/JuliaRegistries/General.git"
       Added registry `General` to `~/.julia/registries/General`
   Installed Hwloc_jll ─────────── v2.2.0+0
   Installed IfElse ────────────── v0.1.0
   Installed OffsetArrays ──────── v1.4.2
   Installed UnPack ────────────── v1.0.2
   Installed Hwloc ─────────────── v1.1.0
   Installed ArrayInterface ────── v2.14.9
   Installed Requires ──────────── v1.1.2
   Installed DocStringExtensions ─ v0.8.3
   Installed Adapt ─────────────── v2.3.0
   Installed SLEEFPirates ──────── v0.6.2
   Installed VectorizationBase ─── v0.14.9
  Downloaded artifact: Hwloc
Updating `~/.julia/environments/v1.7/Project.toml`
  [bdcacae8] + LoopVectorization v0.9.12 `https://github.com/chriselrod/LoopVectorization.jl#master`
Updating `~/.julia/environments/v1.7/Manifest.toml`
  [79e6a3ab] + Adapt v2.3.0
  [4fba245c] + ArrayInterface v2.14.9
  [ffbed154] + DocStringExtensions v0.8.3
  [0e44f5e4] + Hwloc v1.1.0
  [615f187c] + IfElse v0.1.0
  [bdcacae8] + LoopVectorization v0.9.12 `https://github.com/chriselrod/LoopVectorization.jl#master`
  [6fe1bfb0] + OffsetArrays v1.4.2
  [ae029012] + Requires v1.1.2
  [476501e8] + SLEEFPirates v0.6.2
  [3a884ed6] + UnPack v1.0.2
  [3d5dd08c] + VectorizationBase v0.14.9
  [e33a78d0] + Hwloc_jll v2.2.0+0
  [0dad84c5] + ArgTools
  [56f22d72] + Artifacts
  [2a0f44e3] + Base64
  [ade2ca70] + Dates
  [f43a241f] + Downloads
  [b77e0a4c] + InteractiveUtils
  [b27032c2] + LibCURL
  [76f85450] + LibGit2
  [8f399da3] + Libdl
  [37e2e46d] + LinearAlgebra
  [56ddb016] + Logging
  [d6f4376e] + Markdown
  [ca575930] + NetworkOptions
  [44cfe95a] + Pkg
  [de0858da] + Printf
  [3fa0cd96] + REPL
  [9a3f8284] + Random
  [ea8e919c] + SHA
  [9e88b42a] + Serialization
  [6462fe0b] + Sockets
  [2f01184e] + SparseArrays
  [fa267f1f] + TOML
  [a4e569a6] + Tar
  [8dfed614] + Test
  [cf7118a7] + UUIDs
  [4ec0a83e] + Unicode
  [deac9b47] + LibCURL_jll
  [29816b5a] + LibSSH2_jll
  [c8ffd9c3] + MbedTLS_jll
  [14a3606d] + MozillaCACerts_jll
  [83775a58] + Zlib_jll
  [8e850ede] + nghttp2_jll

15 dependencies successfully precompiled in 34 seconds
code.jl.3121.cov
     1	        - using LoopVectorization
     2	        -
     3	        1 function f!(x::AbstractVector)
     4	        2     @avx for i ∈ axes(x, 1)
     5	        0         x[i] = 1
     6	        -     end
     7	        1     return x
     8	        - end
     Cloning git-repo `https://github.com/DilumAluthge/LoopVectorization.jl`
    Updating git-repo `https://github.com/DilumAluthge/LoopVectorization.jl`
   Resolving package versions...
  Installing known registries into `~/.julia`
     Cloning registry from "https://github.com/JuliaRegistries/General.git"
       Added registry `General` to `~/.julia/registries/General`
   Installed UnPack ────────────── v1.0.2
   Installed Hwloc_jll ─────────── v2.2.0+0
   Installed IfElse ────────────── v0.1.0
   Installed Hwloc ─────────────── v1.1.0
   Installed ArrayInterface ────── v2.14.9
   Installed OffsetArrays ──────── v1.4.2
   Installed DocStringExtensions ─ v0.8.3
   Installed Requires ──────────── v1.1.2
   Installed Adapt ─────────────── v2.3.0
   Installed VectorizationBase ─── v0.14.9
   Installed SLEEFPirates ──────── v0.6.2
  Downloaded artifact: Hwloc
Updating `~/.julia/environments/v1.7/Project.toml`
  [bdcacae8] + LoopVectorization v0.9.12 `https://github.com/DilumAluthge/LoopVectorization.jl#dpa/avx-coverage`
Updating `~/.julia/environments/v1.7/Manifest.toml`
  [79e6a3ab] + Adapt v2.3.0
  [4fba245c] + ArrayInterface v2.14.9
  [ffbed154] + DocStringExtensions v0.8.3
  [0e44f5e4] + Hwloc v1.1.0
  [615f187c] + IfElse v0.1.0
  [bdcacae8] + LoopVectorization v0.9.12 `https://github.com/DilumAluthge/LoopVectorization.jl#dpa/avx-coverage`
  [6fe1bfb0] + OffsetArrays v1.4.2
  [ae029012] + Requires v1.1.2
  [476501e8] + SLEEFPirates v0.6.2
  [3a884ed6] + UnPack v1.0.2
  [3d5dd08c] + VectorizationBase v0.14.9
  [e33a78d0] + Hwloc_jll v2.2.0+0
  [0dad84c5] + ArgTools
  [56f22d72] + Artifacts
  [2a0f44e3] + Base64
  [ade2ca70] + Dates
  [f43a241f] + Downloads
  [b77e0a4c] + InteractiveUtils
  [b27032c2] + LibCURL
  [76f85450] + LibGit2
  [8f399da3] + Libdl
  [37e2e46d] + LinearAlgebra
  [56ddb016] + Logging
  [d6f4376e] + Markdown
  [ca575930] + NetworkOptions
  [44cfe95a] + Pkg
  [de0858da] + Printf
  [3fa0cd96] + REPL
  [9a3f8284] + Random
  [ea8e919c] + SHA
  [9e88b42a] + Serialization
  [6462fe0b] + Sockets
  [2f01184e] + SparseArrays
  [fa267f1f] + TOML
  [a4e569a6] + Tar
  [8dfed614] + Test
  [cf7118a7] + UUIDs
  [4ec0a83e] + Unicode
  [deac9b47] + LibCURL_jll
  [29816b5a] + LibSSH2_jll
  [c8ffd9c3] + MbedTLS_jll
  [14a3606d] + MozillaCACerts_jll
  [83775a58] + Zlib_jll
  [8e850ede] + nghttp2_jll

15 dependencies successfully precompiled in 26 seconds
code.jl.3281.cov
     1	        - using LoopVectorization
     2	        -
     3	        1 function f!(x::AbstractVector)
     4	        1     @avx for i ∈ axes(x, 1)
     5	        2         x[i] = 1
     6	        -     end
     7	        1     return x
     8	        - end
```

In particular, note that with `LoopVectorization#master`, we have:
```
code.jl.3121.cov
     1	        - using LoopVectorization
     2	        -
     3	        1 function f!(x::AbstractVector)
     4	        2     @avx for i ∈ axes(x, 1)
     5	        0         x[i] = 1
     6	        -     end
     7	        1     return x
     8	        - end
```

And on this PR branch, we have:
```
code.jl.3281.cov
     1	        - using LoopVectorization
     2	        -
     3	        1 function f!(x::AbstractVector)
     4	        1     @avx for i ∈ axes(x, 1)
     5	        2         x[i] = 1
     6	        -     end
     7	        1     return x
     8	        - end
```

DilumAluthge commented:
When you do get around to the overhaul, I'm happy to help out with the line number stuff.

DilumAluthge commented:
This doesn't seem to be sufficient. E.g. take a look at the Gaius.jl PR.

chriselrod commented:
Yeah, it only gets 1 of the lines. What do the dumped nodes look like? Maybe they need to be interleaved.

DilumAluthge commented:
I think Julia expects that each `LineNumberNode` is associated with at least one actual line of code.

So I put in some dummy `nothing`s.

It works for me when I try one of the Gaius kernels locally. Let's see if it fixes Gaius.

DilumAluthge commented:
This is what I get locally:
```
        - import LoopVectorization
        - 
        1 function gemm_kernel!(C, A, B)
        1     LoopVectorization.@avx for n ∈ 1:size(A, 1), m ∈ 1:size(B, 2)
        1         Cₙₘ = zero(eltype(C))
        1         for k ∈ 1:size(A, 2)
        1             Cₙₘ += A[n,k] * B[k,m]
        -         end
        1         C[n,m] = Cₙₘ
        -     end
        - end

```

DilumAluthge commented:
I think it's working now. See e.g. https://app.codecov.io/gh/MasonProtter/Gaius.jl/compare/45/changes

DilumAluthge commented:
@chriselrod Is there any chance that inserting these `nothing`s would have a negative impact on runtime performance?

DilumAluthge commented:
I tried replacing the `:(nothing)` with `Expr(:block)`, but that doesn't work.

So far, the only thing that I've tried that works is putting a `:(nothing)` after each `LineNumberNode`.

chriselrod commented:
> @chriselrod Is there any chance that inserting these `nothing`s would have a negative impact on runtime performance?

I seriously doubt it. The compiler should strip all of them / there is no code for them to actually get converted to.
But I can compare a few `@code_native`s before/after to be sure.

DilumAluthge commented:
> > @chriselrod Is there any chance that inserting these `nothing`s would have a negative impact on runtime performance?
> 
> I seriously doubt it. The compiler should strip all of them / there is no code for them to actually get converted to.
> But I can compare a few `@code_native`s before/after to be sure.

Yeah that would be great, just to put my mind at ease 😂

chriselrod commented:
Actually, `@code_typed` or the default view from `@descend` are a lot nicer here.
First, as can see all the line number nodes:
```julia
julia> @macroexpand @avx for m ∈ axes(A,1), n ∈ axes(B,2)
               ΔCₘₙ = zero(eltype(C))
               for k ∈ axes(A,2)
                   ΔCₘₙ += A[m,k] * B[k,n]
               end
               C[m,n] = ΔCₘₙ
           end
quote
    begin
        #= REPL[10]:6 =#
        nothing
    end
    begin
        #= REPL[10]:4 =#
        nothing
    end
    begin
        #= REPL[10]:3 =#
        nothing
    end
    begin
        #= REPL[10]:2 =#
        nothing
    end
    begin
        var"###loopm###1###" = LoopVectorization.maybestaticrange(axes(A, LoopVectorization.Static{1}()))
        var"###m_loop_lower_bound###2###" = LoopVectorization.maybestaticfirst(var"###loopm###1###")
        var"###m_loop_upper_bound###3###" = LoopVectorization.maybestaticlast(var"###loopm###1###")
        var"###loopn###4###" = LoopVectorization.maybestaticrange(axes(B, LoopVectorization.Static{2}()))
        var"###n_loop_lower_bound###5###" = LoopVectorization.maybestaticfirst(var"###loopn###4###")
        var"###n_loop_upper_bound###6###" = LoopVectorization.maybestaticlast(var"###loopn###4###")
        var"###loopk###8###" = LoopVectorization.maybestaticrange(axes(A, LoopVectorization.Static{2}()))
        var"###k_loop_lower_bound###9###" = LoopVectorization.maybestaticfirst(var"###loopk###8###")
        var"###k_loop_upper_bound###10###" = LoopVectorization.maybestaticlast(var"###loopk###8###")
    end
    if LoopVectorization.check_args(A, B, C)
        var"##vptr##_A" = LoopVectorization.stridedpointer(A)
        var"##vptr##_B" = LoopVectorization.stridedpointer(B)
        var"##vptr##_C" = LoopVectorization.stridedpointer(C)
        begin
            var"##A#468" = LoopVectorization.preserve_buffer(A)
            var"##B#469" = LoopVectorization.preserve_buffer(B)
            var"##C#470" = LoopVectorization.preserve_buffer(C)
            $(Expr(:gc_preserve, quote
    begin
        var"##Tloopeltype##" = LoopVectorization.promote_type(LoopVectorization.eltype(A), LoopVectorization.eltype(B), LoopVectorization.eltype(C))
        var"##Wvecwidth##" = LoopVectorization.pick_vector_width_val(LoopVectorization._eltype(A), LoopVectorization._eltype(B), LoopVectorization._eltype(C))
    end
    LoopVectorization._avx_!(Val{(0, 0, 0, LoopVectorization.unwrap(var"##Wvecwidth##"))}(), Tuple{:numericconstant, Symbol("###zero###7###"), LoopVectorization.OperationStruct(0x0000000000000012, 0x0000000000000000, 0x0000000000000003, 0x0000000000000000, LoopVectorization.constant, 0x00, 0x01), :LoopVectorization, :getindex, LoopVectorization.OperationStruct(0x0000000000000013, 0x0000000000000000, 0x0000000000000000, 0x0000000000000000, LoopVectorization.memload, 0x01, 0x02), :LoopVectorization, :getindex, LoopVectorization.OperationStruct(0x0000000000000032, 0x0000000000000000, 0x0000000000000000, 0x0000000000000000, LoopVectorization.memload, 0x02, 0x03), :LoopVectorization, :vfmadd, LoopVectorization.OperationStruct(0x0000000000000132, 0x0000000000000003, 0x0000000000000000, 0x0000000000020301, LoopVectorization.compute, 0x00, 0x01), :LoopVectorization, :identity, LoopVectorization.OperationStruct(0x0000000000000012, 0x0000000000000003, 0x0000000000000000, 0x0000000000000004, LoopVectorization.compute, 0x00, 0x01), :LoopVectorization, :setindex!, LoopVectorization.OperationStruct(0x0000000000000012, 0x0000000000000003, 0x0000000000000000, 0x0000000000000005, LoopVectorization.memstore, 0x03, 0x04)}, Tuple{LoopVectorization.ArrayRefStruct{:A, Symbol("##vptr##_A")}(0x0000000000000101, 0x0000000000000103, 0x0000000000000000), LoopVectorization.ArrayRefStruct{:B, Symbol("##vptr##_B")}(0x0000000000000101, 0x0000000000000302, 0x0000000000000000), LoopVectorization.ArrayRefStruct{:C, Symbol("##vptr##_C")}(0x0000000000000101, 0x0000000000000102, 0x0000000000000000)}, Tuple{0, Tuple{}, Tuple{}, Tuple{}, Tuple{}, Tuple{(1, LoopVectorization.IntOrFloat)}, Tuple{}}, Tuple{:m, :n, :k}, (var"###m_loop_lower_bound###2###":var"###m_loop_upper_bound###3###", var"###n_loop_lower_bound###5###":var"###n_loop_upper_bound###6###", var"###k_loop_lower_bound###9###":var"###k_loop_upper_bound###10###"), var"##vptr##_A", var"##vptr##_B", var"##vptr##_C")
    nothing
end, Symbol("##A#468"), Symbol("##B#469"), Symbol("##C#470")))
        end
    else
        $(Expr(:inbounds, true))
        local var"#3044#val" = for m = axes(A, 1), n = axes(B, 2)
                    #= REPL[10]:2 =#
                    ΔCₘₙ = zero(eltype(C))
                    #= REPL[10]:3 =#
                    for k = axes(A, 2)
                        #= REPL[10]:4 =#
                        ΔCₘₙ = Base.FastMath.add_fast(ΔCₘₙ, Base.FastMath.mul_fast(A[m, k], B[k, n]))
                    end
                    #= REPL[10]:6 =#
                    C[m, n] = ΔCₘₙ
                end
        $(Expr(:inbounds, :pop))
        var"#3044#val"
    end
end
```
But:
```julia
julia> function AmulB!(C, A, B)
           @avx for m ∈ axes(A,1), n ∈ axes(B,2)
               ΔCₘₙ = zero(eltype(C))
               for k ∈ axes(A,2)
                   ΔCₘₙ += A[m,k] * B[k,n]
               end
               C[m,n] = ΔCₘₙ
           end
       end

julia> M=K=N=200; A = rand(M,K); B = rand(K,N); C1 = Matrix{Float64}(undef, M, N);

julia> @code_typed AmulB!(C1,A,B)
CodeInfo(
1 ─ %1  = Base.arraysize(A, 1)::Int64
│         Base.arraysize(A, 2)::Int64
│   %3  = Base.slt_int(%1, 0)::Bool
│   %4  = Base.ifelse(%3, 0, %1)::Int64
│         Base.arraysize(B, 1)::Int64
│   %6  = Base.arraysize(B, 2)::Int64
│   %7  = Base.slt_int(%6, 0)::Bool
│   %8  = Base.ifelse(%7, 0, %6)::Int64
│         Base.arraysize(A, 1)::Int64
│   %10 = Base.arraysize(A, 2)::Int64
│   %11 = Base.slt_int(%10, 0)::Bool
│   %12 = Base.ifelse(%11, 0, %10)::Int64
│   %13 = $(Expr(:foreigncall, :(:jl_array_ptr), Ptr{Float64}, svec(Any), 0, :(:ccall), Core.Argument(3)))::Ptr{Float64}
│   %14 = Base.arraysize(A, 1)::Int64
│         Base.arraysize(A, 2)::Int64
│   %16 = Base.mul_int(1, %14)::Int64
│   %17 = VectorizationBase.llvmcall("%res = mul nsw i64 %0, %1\nret i64 %res", Int64, Tuple{Int64, Int64}, 8, %16)::Int64
│   %18 = Core.tuple(Static(8), %17)::Tuple{ArrayInterface.StaticInt{8}, Int64}
│         Base.arraysize(A, 1)::Int64
│         Base.arraysize(A, 2)::Int64
│         Base.arraysize(A, 1)::Int64
│         Base.arraysize(A, 2)::Int64
│   %23 = %new(VectorizationBase.StridedPointer{Float64, 2, 1, 0, (1, 2), Tuple{ArrayInterface.StaticInt{8}, Int64}, Tuple{ArrayInterface.StaticInt{1}, ArrayInterface.StaticInt{1}}}, %13, %18, (Static(1), Static(1)))::VectorizationBase.StridedPointer{Float64, 2, 1, 0, (1, 2), Tuple{ArrayInterface.StaticInt{8}, Int64}, Tuple{ArrayInterface.StaticInt{1}, ArrayInterface.StaticInt{1}}}
│   %24 = $(Expr(:foreigncall, :(:jl_array_ptr), Ptr{Float64}, svec(Any), 0, :(:ccall), Core.Argument(4)))::Ptr{Float64}
│   %25 = Base.arraysize(B, 1)::Int64
│         Base.arraysize(B, 2)::Int64
│   %27 = Base.mul_int(1, %25)::Int64
│   %28 = VectorizationBase.llvmcall("%res = mul nsw i64 %0, %1\nret i64 %res", Int64, Tuple{Int64, Int64}, 8, %27)::Int64
│   %29 = Core.tuple(Static(8), %28)::Tuple{ArrayInterface.StaticInt{8}, Int64}
│         Base.arraysize(B, 1)::Int64
│         Base.arraysize(B, 2)::Int64
│         Base.arraysize(B, 1)::Int64
│         Base.arraysize(B, 2)::Int64
│   %34 = %new(VectorizationBase.StridedPointer{Float64, 2, 1, 0, (1, 2), Tuple{ArrayInterface.StaticInt{8}, Int64}, Tuple{ArrayInterface.StaticInt{1}, ArrayInterface.StaticInt{1}}}, %24, %29, (Static(1), Static(1)))::VectorizationBase.StridedPointer{Float64, 2, 1, 0, (1, 2), Tuple{ArrayInterface.StaticInt{8}, Int64}, Tuple{ArrayInterface.StaticInt{1}, ArrayInterface.StaticInt{1}}}
│   %35 = $(Expr(:foreigncall, :(:jl_array_ptr), Ptr{Float64}, svec(Any), 0, :(:ccall), Core.Argument(2)))::Ptr{Float64}
│   %36 = Base.arraysize(C, 1)::Int64
│         Base.arraysize(C, 2)::Int64
│   %38 = Base.mul_int(1, %36)::Int64
│   %39 = VectorizationBase.llvmcall("%res = mul nsw i64 %0, %1\nret i64 %res", Int64, Tuple{Int64, Int64}, 8, %38)::Int64
│   %40 = Core.tuple(Static(8), %39)::Tuple{ArrayInterface.StaticInt{8}, Int64}
│         Base.arraysize(C, 1)::Int64
│         Base.arraysize(C, 2)::Int64
│         Base.arraysize(C, 1)::Int64
│         Base.arraysize(C, 2)::Int64
│   %45 = %new(VectorizationBase.StridedPointer{Float64, 2, 1, 0, (1, 2), Tuple{ArrayInterface.StaticInt{8}, Int64}, Tuple{ArrayInterface.StaticInt{1}, ArrayInterface.StaticInt{1}}}, %35, %40, (Static(1), Static(1)))::VectorizationBase.StridedPointer{Float64, 2, 1, 0, (1, 2), Tuple{ArrayInterface.StaticInt{8}, Int64}, Tuple{ArrayInterface.StaticInt{1}, ArrayInterface.StaticInt{1}}}
│   %46 = $(Expr(:gc_preserve_begin, Core.Argument(3), Core.Argument(4), Core.Argument(2)))
│   %47 = %new(ArrayInterface.OptionallyStaticUnitRange{ArrayInterface.StaticInt{1}, Int64}, Static(1), %4)::ArrayInterface.OptionallyStaticUnitRange{ArrayInterface.StaticInt{1}, Int64}
│   %48 = %new(ArrayInterface.OptionallyStaticUnitRange{ArrayInterface.StaticInt{1}, Int64}, Static(1), %8)::ArrayInterface.OptionallyStaticUnitRange{ArrayInterface.StaticInt{1}, Int64}
│   %49 = %new(ArrayInterface.OptionallyStaticUnitRange{ArrayInterface.StaticInt{1}, Int64}, Static(1), %12)::ArrayInterface.OptionallyStaticUnitRange{ArrayInterface.StaticInt{1}, Int64}
│   %50 = Core.tuple(%47, %48, %49)::Tuple{ArrayInterface.OptionallyStaticUnitRange{ArrayInterface.StaticInt{1}, Int64}, ArrayInterface.OptionallyStaticUnitRange{ArrayInterface.StaticInt{1}, Int64}, ArrayInterface.OptionallyStaticUnitRange{ArrayInterface.StaticInt{1}, Int64}}
│         invoke LoopVectorization._avx_!($(QuoteNode(Val{(0, 0, 0, 8)}()))::Val{(0, 0, 0, 8)}, Tuple{:numericconstant, Symbol("###zero###7###"), LoopVectorization.OperationStruct(0x0000000000000012, 0x0000000000000000, 0x0000000000000003, 0x0000000000000000, LoopVectorization.constant, 0x00, 0x01), :LoopVectorization, :getindex, LoopVectorization.OperationStruct(0x0000000000000013, 0x0000000000000000, 0x0000000000000000, 0x0000000000000000, LoopVectorization.memload, 0x01, 0x02), :LoopVectorization, :getindex, LoopVectorization.OperationStruct(0x0000000000000032, 0x0000000000000000, 0x0000000000000000, 0x0000000000000000, LoopVectorization.memload, 0x02, 0x03), :LoopVectorization, :vfmadd, LoopVectorization.OperationStruct(0x0000000000000132, 0x0000000000000003, 0x0000000000000000, 0x0000000000020301, LoopVectorization.compute, 0x00, 0x01), :LoopVectorization, :identity, LoopVectorization.OperationStruct(0x0000000000000012, 0x0000000000000003, 0x0000000000000000, 0x0000000000000004, LoopVectorization.compute, 0x00, 0x01), :LoopVectorization, :setindex!, LoopVectorization.OperationStruct(0x0000000000000012, 0x0000000000000003, 0x0000000000000000, 0x0000000000000005, LoopVectorization.memstore, 0x03, 0x04)}::Type{Tuple{:numericconstant, Symbol("###zero###7###"), LoopVectorization.OperationStruct(0x0000000000000012, 0x0000000000000000, 0x0000000000000003, 0x0000000000000000, LoopVectorization.constant, 0x00, 0x01), :LoopVectorization, :getindex, LoopVectorization.OperationStruct(0x0000000000000013, 0x0000000000000000, 0x0000000000000000, 0x0000000000000000, LoopVectorization.memload, 0x01, 0x02), :LoopVectorization, :getindex, LoopVectorization.OperationStruct(0x0000000000000032, 0x0000000000000000, 0x0000000000000000, 0x0000000000000000, LoopVectorization.memload, 0x02, 0x03), :LoopVectorization, :vfmadd, LoopVectorization.OperationStruct(0x0000000000000132, 0x0000000000000003, 0x0000000000000000, 0x0000000000020301, LoopVectorization.compute, 0x00, 0x01), :LoopVectorization, :identity, LoopVectorization.OperationStruct(0x0000000000000012, 0x0000000000000003, 0x0000000000000000, 0x0000000000000004, LoopVectorization.compute, 0x00, 0x01), :LoopVectorization, :setindex!, LoopVectorization.OperationStruct(0x0000000000000012, 0x0000000000000003, 0x0000000000000000, 0x0000000000000005, LoopVectorization.memstore, 0x03, 0x04)}}, Tuple{LoopVectorization.ArrayRefStruct{:A, Symbol("##vptr##_A")}(0x0000000000000101, 0x0000000000000103, 0x0000000000000000), LoopVectorization.ArrayRefStruct{:B, Symbol("##vptr##_B")}(0x0000000000000101, 0x0000000000000302, 0x0000000000000000), LoopVectorization.ArrayRefStruct{:C, Symbol("##vptr##_C")}(0x0000000000000101, 0x0000000000000102, 0x0000000000000000)}::Type{Tuple{LoopVectorization.ArrayRefStruct{:A, Symbol("##vptr##_A")}(0x0000000000000101, 0x0000000000000103, 0x0000000000000000), LoopVectorization.ArrayRefStruct{:B, Symbol("##vptr##_B")}(0x0000000000000101, 0x0000000000000302, 0x0000000000000000), LoopVectorization.ArrayRefStruct{:C, Symbol("##vptr##_C")}(0x0000000000000101, 0x0000000000000102, 0x0000000000000000)}}, Tuple{0, Tuple{}, Tuple{}, Tuple{}, Tuple{}, Tuple{(1, LoopVectorization.IntOrFloat)}, Tuple{}}::Type{Tuple{0, Tuple{}, Tuple{}, Tuple{}, Tuple{}, Tuple{(1, LoopVectorization.IntOrFloat)}, Tuple{}}}, Tuple{:m, :n, :k}::Type{Tuple{:m, :n, :k}}, %50::Tuple{ArrayInterface.OptionallyStaticUnitRange{ArrayInterface.StaticInt{1}, Int64}, ArrayInterface.OptionallyStaticUnitRange{ArrayInterface.StaticInt{1}, Int64}, ArrayInterface.OptionallyStaticUnitRange{ArrayInterface.StaticInt{1}, Int64}}, %23::VectorizationBase.StridedPointer{Float64, 2, 1, 0, (1, 2), Tuple{ArrayInterface.StaticInt{8}, Int64}, Tuple{ArrayInterface.StaticInt{1}, ArrayInterface.StaticInt{1}}}, %34::VectorizationBase.StridedPointer{Float64, 2, 1, 0, (1, 2), Tuple{ArrayInterface.StaticInt{8}, Int64}, Tuple{ArrayInterface.StaticInt{1}, ArrayInterface.StaticInt{1}}}, %45::VectorizationBase.StridedPointer{Float64, 2, 1, 0, (1, 2), Tuple{ArrayInterface.StaticInt{8}, Int64}, Tuple{ArrayInterface.StaticInt{1}, ArrayInterface.StaticInt{1}}})::Any
│         $(Expr(:gc_preserve_end, :(%46)))
└──       return nothing
) => Nothing

```

DilumAluthge commented:
You beat me to it!

I ran this script on `LoopVectorization#6987acf7` versus this PR branch:
```julia
using InteractiveUtils
using LoopVectorization

function gemm_kernel!(C, A, B)
    @avx for n ∈ 1:size(A, 1), m ∈ 1:size(B, 2)
        Cₙₘ = zero(eltype(C))
        for k ∈ 1:size(A, 2)
            Cₙₘ += A[n,k] * B[k,m]
        end
        C[n,m] = Cₙₘ
    end
end

A = rand(Float64, 100, 200)
B = randn(Float64, 200, 300)
C = Matrix{Float64}(undef, 100, 300)
@code_native gemm_kernel!(C, A, B)
```

Results on `LoopVectorization#6987acf7`:
```assembly
	.section	__TEXT,__text,regular,pure_instructions
; ┌ @ script.jl:4 within `gemm_kernel!'
	pushq	%rbp
	movq	%rsp, %rbp
	pushq	%r14
	pushq	%rbx
	andq	$-32, %rsp
	subq	$128, %rsp
	movq	%rsi, %rbx
	vxorps	%xmm0, %xmm0, %xmm0
	vmovaps	%ymm0, (%rsp)
	movq	$0, 32(%rsp)
	movq	%rsi, 120(%rsp)
	movabsq	$jl_get_ptls_states_fast, %rax
	vzeroupper
	callq	*%rax
	movq	%rax, %r14
	movq	$12, (%rsp)
	movq	(%rax), %rax
	movq	%rax, 8(%rsp)
	movq	%rsp, %rax
	movq	%rax, (%r14)
	movq	(%rbx), %rax
	movq	8(%rbx), %rcx
	movq	16(%rbx), %rdx
; │ @ script.jl:5 within `gemm_kernel!'
; │┌ @ stridedpointers.jl:72 within `stridedpointer'
; ││┌ @ stridedpointers.jl:35 within `memory_reference' @ stridedpointers.jl:36
; │││┌ @ abstractarray.jl:1110 within `pointer'
; ││││┌ @ pointer.jl:65 within `unsafe_convert'
	movq	(%rcx), %rsi
; │└└└└
; │┌ @ static.jl:47 within `maybestaticsize'
; ││┌ @ dimensions.jl:155 within `size' @ array.jl:134
	movq	24(%rcx), %rdi
	movq	32(%rdx), %r9
	movq	32(%rcx), %r8
; │└└
; │┌ @ stridedpointers.jl:72 within `stridedpointer'
; ││┌ @ stridedpointers.jl:21 within `bytestrides'
; │││┌ @ stridedpointers.jl:19 within `mulsizeof' @ stridedpointers.jl:18 @ stridedpointers.jl:15
; ││││┌ @ binary_ops.jl:48 within `vmul'
; │││││┌ @ binary_ops.jl:45 within `_vmul'
; ││││││┌ @ binary_ops.jl:45 within `macro expansion'
	leaq	(,%rdi,8), %rbx
; ││└└└└└
; ││ @ stridedpointers.jl:72 within `stridedpointer' @ stridedpointers.jl:77
; ││┌ @ stridedpointers.jl:64 within `StridedPointer'
	movq	%rsi, 80(%rsp)
	movq	%rbx, 88(%rsp)
; ││└
; ││ @ stridedpointers.jl:72 within `stridedpointer'
; ││┌ @ stridedpointers.jl:35 within `memory_reference' @ stridedpointers.jl:36
; │││┌ @ abstractarray.jl:1110 within `pointer'
; ││││┌ @ pointer.jl:65 within `unsafe_convert'
	movq	(%rdx), %rsi
; ││└└└
; ││┌ @ stridedpointers.jl:21 within `bytestrides'
; │││┌ @ stridelayout.jl:561 within `strides' @ reinterpretarray.jl:160
; ││││┌ @ array.jl:134 within `size'
	movq	24(%rdx), %rbx
; │││└└
; │││┌ @ stridedpointers.jl:19 within `mulsizeof' @ stridedpointers.jl:18 @ stridedpointers.jl:15
; ││││┌ @ binary_ops.jl:48 within `vmul'
; │││││┌ @ binary_ops.jl:45 within `_vmul'
; ││││││┌ @ binary_ops.jl:45 within `macro expansion'
	shlq	$3, %rbx
; ││└└└└└
; ││ @ stridedpointers.jl:72 within `stridedpointer' @ stridedpointers.jl:77
; ││┌ @ stridedpointers.jl:64 within `StridedPointer'
	movq	%rsi, 64(%rsp)
	movq	%rbx, 72(%rsp)
; ││└
; ││ @ stridedpointers.jl:72 within `stridedpointer'
; ││┌ @ stridedpointers.jl:35 within `memory_reference' @ stridedpointers.jl:36
; │││┌ @ abstractarray.jl:1110 within `pointer'
; ││││┌ @ pointer.jl:65 within `unsafe_convert'
	movq	(%rax), %rsi
; ││└└└
; ││┌ @ stridedpointers.jl:21 within `bytestrides'
; │││┌ @ stridelayout.jl:561 within `strides' @ reinterpretarray.jl:160
; ││││┌ @ array.jl:134 within `size'
	movq	24(%rax), %rbx
; │││└└
; │││┌ @ stridedpointers.jl:19 within `mulsizeof' @ stridedpointers.jl:18 @ stridedpointers.jl:15
; ││││┌ @ binary_ops.jl:48 within `vmul'
; │││││┌ @ binary_ops.jl:45 within `_vmul'
; ││││││┌ @ binary_ops.jl:45 within `macro expansion'
	shlq	$3, %rbx
; ││└└└└└
; ││ @ stridedpointers.jl:72 within `stridedpointer' @ stridedpointers.jl:77
; ││┌ @ stridedpointers.jl:64 within `StridedPointer'
	movq	%rsi, 48(%rsp)
	movq	%rbx, 56(%rsp)
; │└└
	movq	%rdi, 96(%rsp)
	movq	%r9, 104(%rsp)
	movq	%r8, 112(%rsp)
	movq	%rcx, 32(%rsp)
	movq	%rdx, 24(%rsp)
	movq	%rax, 16(%rsp)
	leaq	48(%rsp), %rax
	leaq	64(%rsp), %rbx
	movabsq	$"_avx_!", %r10
	movabsq	$4772200016, %rdi               ## imm = 0x11C71FE50
	movabsq	$4772200112, %rsi               ## imm = 0x11C71FEB0
	movabsq	$4772053968, %rdx               ## imm = 0x11C6FC3D0
	movabsq	$4772054064, %rcx               ## imm = 0x11C6FC430
	leaq	96(%rsp), %r8
	leaq	80(%rsp), %r9
	pushq	%rax
	pushq	%rbx
	callq	*%r10
	addq	$16, %rsp
	movq	8(%rsp), %rax
	movq	%rax, (%r14)
	movabsq	$jl_system_image_data, %rax
	leaq	-16(%rbp), %rsp
	popq	%rbx
	popq	%r14
	popq	%rbp
	retq
; └

```

Results on the PR branch:
```assembly
	.section	__TEXT,__text,regular,pure_instructions
; ┌ @ script.jl:4 within `gemm_kernel!'
	pushq	%rbp
	movq	%rsp, %rbp
	pushq	%r14
	pushq	%rbx
	andq	$-32, %rsp
	subq	$128, %rsp
	movq	%rsi, %rbx
	vxorps	%xmm0, %xmm0, %xmm0
	vmovaps	%ymm0, (%rsp)
	movq	$0, 32(%rsp)
	movq	%rsi, 120(%rsp)
	movabsq	$jl_get_ptls_states_fast, %rax
	vzeroupper
	callq	*%rax
	movq	%rax, %r14
	movq	$12, (%rsp)
	movq	(%rax), %rax
	movq	%rax, 8(%rsp)
	movq	%rsp, %rax
	movq	%rax, (%r14)
	movq	(%rbx), %rax
	movq	8(%rbx), %rcx
	movq	16(%rbx), %rdx
; │ @ script.jl:6 within `gemm_kernel!'
; │┌ @ stridedpointers.jl:72 within `stridedpointer'
; ││┌ @ stridedpointers.jl:35 within `memory_reference' @ stridedpointers.jl:36
; │││┌ @ abstractarray.jl:1110 within `pointer'
; ││││┌ @ pointer.jl:65 within `unsafe_convert'
	movq	(%rcx), %rsi
; │└└└└
; │┌ @ static.jl:47 within `maybestaticsize'
; ││┌ @ dimensions.jl:155 within `size' @ array.jl:134
	movq	24(%rcx), %rdi
	movq	32(%rdx), %r9
	movq	32(%rcx), %r8
; │└└
; │┌ @ stridedpointers.jl:72 within `stridedpointer'
; ││┌ @ stridedpointers.jl:21 within `bytestrides'
; │││┌ @ stridedpointers.jl:19 within `mulsizeof' @ stridedpointers.jl:18 @ stridedpointers.jl:15
; ││││┌ @ binary_ops.jl:48 within `vmul'
; │││││┌ @ binary_ops.jl:45 within `_vmul'
; ││││││┌ @ binary_ops.jl:45 within `macro expansion'
	leaq	(,%rdi,8), %rbx
; ││└└└└└
; ││ @ stridedpointers.jl:72 within `stridedpointer' @ stridedpointers.jl:77
; ││┌ @ stridedpointers.jl:64 within `StridedPointer'
	movq	%rsi, 80(%rsp)
	movq	%rbx, 88(%rsp)
; ││└
; ││ @ stridedpointers.jl:72 within `stridedpointer'
; ││┌ @ stridedpointers.jl:35 within `memory_reference' @ stridedpointers.jl:36
; │││┌ @ abstractarray.jl:1110 within `pointer'
; ││││┌ @ pointer.jl:65 within `unsafe_convert'
	movq	(%rdx), %rsi
; ││└└└
; ││┌ @ stridedpointers.jl:21 within `bytestrides'
; │││┌ @ stridelayout.jl:561 within `strides' @ reinterpretarray.jl:160
; ││││┌ @ array.jl:134 within `size'
	movq	24(%rdx), %rbx
; │││└└
; │││┌ @ stridedpointers.jl:19 within `mulsizeof' @ stridedpointers.jl:18 @ stridedpointers.jl:15
; ││││┌ @ binary_ops.jl:48 within `vmul'
; │││││┌ @ binary_ops.jl:45 within `_vmul'
; ││││││┌ @ binary_ops.jl:45 within `macro expansion'
	shlq	$3, %rbx
; ││└└└└└
; ││ @ stridedpointers.jl:72 within `stridedpointer' @ stridedpointers.jl:77
; ││┌ @ stridedpointers.jl:64 within `StridedPointer'
	movq	%rsi, 64(%rsp)
	movq	%rbx, 72(%rsp)
; ││└
; ││ @ stridedpointers.jl:72 within `stridedpointer'
; ││┌ @ stridedpointers.jl:35 within `memory_reference' @ stridedpointers.jl:36
; │││┌ @ abstractarray.jl:1110 within `pointer'
; ││││┌ @ pointer.jl:65 within `unsafe_convert'
	movq	(%rax), %rsi
; ││└└└
; ││┌ @ stridedpointers.jl:21 within `bytestrides'
; │││┌ @ stridelayout.jl:561 within `strides' @ reinterpretarray.jl:160
; ││││┌ @ array.jl:134 within `size'
	movq	24(%rax), %rbx
; │││└└
; │││┌ @ stridedpointers.jl:19 within `mulsizeof' @ stridedpointers.jl:18 @ stridedpointers.jl:15
; ││││┌ @ binary_ops.jl:48 within `vmul'
; │││││┌ @ binary_ops.jl:45 within `_vmul'
; ││││││┌ @ binary_ops.jl:45 within `macro expansion'
	shlq	$3, %rbx
; ││└└└└└
; ││ @ stridedpointers.jl:72 within `stridedpointer' @ stridedpointers.jl:77
; ││┌ @ stridedpointers.jl:64 within `StridedPointer'
	movq	%rsi, 48(%rsp)
	movq	%rbx, 56(%rsp)
; │└└
	movq	%rdi, 96(%rsp)
	movq	%r9, 104(%rsp)
	movq	%r8, 112(%rsp)
	movq	%rcx, 32(%rsp)
	movq	%rdx, 24(%rsp)
	movq	%rax, 16(%rsp)
	leaq	48(%rsp), %rax
	leaq	64(%rsp), %rbx
	movabsq	$"_avx_!", %r10
	movabsq	$4571587312, %rdi               ## imm = 0x1107CE2F0
	movabsq	$4571587408, %rsi               ## imm = 0x1107CE350
	movabsq	$4571588656, %rdx               ## imm = 0x1107CE830
	movabsq	$4571588752, %rcx               ## imm = 0x1107CE890
	leaq	96(%rsp), %r8
	leaq	80(%rsp), %r9
	pushq	%rax
	pushq	%rbx
	callq	*%r10
	addq	$16, %rsp
	movq	8(%rsp), %rax
	movq	%rax, (%r14)
	movabsq	$jl_system_image_data, %rax
	leaq	-16(%rbp), %rsp
	popq	%rbx
	popq	%r14
	popq	%rbp
	retq
; └
```

This is the diff:
```diff
26c26
< ; │ @ script.jl:5 within `gemm_kernel!'
---
> ; │ @ script.jl:6 within `gemm_kernel!'
105,108c105,108
< 	movabsq	$4772200016, %rdi               ## imm = 0x11C71FE50
< 	movabsq	$4772200112, %rsi               ## imm = 0x11C71FEB0
< 	movabsq	$4772053968, %rdx               ## imm = 0x11C6FC3D0
< 	movabsq	$4772054064, %rcx               ## imm = 0x11C6FC430
---
> 	movabsq	$4571587312, %rdi               ## imm = 0x1107CE2F0
> 	movabsq	$4571587408, %rsi               ## imm = 0x1107CE350
> 	movabsq	$4571588656, %rdx               ## imm = 0x1107CE830
> 	movabsq	$4571588752, %rcx               ## imm = 0x1107CE890
```

I know virtually nothing about assembly language, but I'm assuming that this difference is not important?
